### PR TITLE
TINKERPOP-1734 DSL for Gremlin .NET

### PR DESCRIPTION
**Pull request by spmallette**
https://issues.apache.org/jira/browse/TINKERPOP-1734

Provided tests and documentation for implementing DSLs in .Net. No changes were needed to the code base for this - just used C# extension methods as suggested by @jorgebay . Only downside was anonymous traversals are a little weird because C# doesn't let you extend static classes and extensions methods can't be used on them either. The only recourse is to wrap static methods in the extending class which is a lot of extra boilerplate. I opted to just recommend that users create their own class for spawning anonymous traversals that just works independently of `__`. 

VOTE +1

### CODE REVIEW

**Comment by FlorianHockmann** [comment]
Really nice to see how easy it is to implement a DSL in Gremlin.Net and I like your suggestion to simply create another version of `__` to deal with the fact that we can't extend it in C#.

My only suggestion is to change the name of the classes in the examples that contain extension methods to end on `Extensions` as that is a convention in C#. So `SocialTraversal` would become `SocialTraversalExtensions` and `SocialTraversalSource` becomes `SocialTraversalSourceExtensions`. That also communicates the fact that those classes don't inherit from `GraphTraversal` and `GraphTraversalSource`.

File: docs/src/reference/the-traversal.asciidoc [hunk 1]
```diff
@@ -3218,3 +3218,93 @@ social.persons("marko").youngestFriendsAge()
...
+}
+----
+
+Note the creation of `__Social` as the Social DSL's "extension" to the available ways in which to spawn anonymous
+traversals. The use of the double underscore prefix in the name is just a convention to consider using and is not a
+requirement. To use the DSL, bring it into scope with the `using` directive:
+
+[source,csharp]
+----
+using Dsl;
```
Comment (jorgebay) [inline docs/src/reference/the-traversal.asciidoc]: I would include a [`using static` directive](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/using-static) in the code sample:

```csharp
using static Dsl.__Social;
```

which allows using `__Social` methods directly:

```csharp
social.Persons().Filter(CreatedAtLeast(2)).Count();
```

**Comment by spmallette** [comment]
Addressed all the feedback with the latest changed and rebased to get rid of the changelog conflict.

**Comment by jorgebay** [comment]
VOTE +1

**Comment by FlorianHockmann** [comment]
VOTE +1

**Pull request merged**